src/sync/autostopper.js

```diff
@@ -0,0 +1,125 @@
+import { ExhaustivenessError, assert } from '../utils.js'
+
+/**
+ * @internal
+ * @typedef {(
+ *   | { type: 'incomplete' }
+ *   | {
+ *     type: 'done';
+ *     doneAt: number;
+ *     timeoutId: null | ReturnType<typeof setTimeout>;
+ *   }
+ *   | { type: 'stopped' }
+ * )} AutostopperState
+ */
+
+/**
+ * `Autostopper` manages auto-stop state for sync.
+ *
+ * Tell `Autostopper` the current state of sync (is it done?) and your desired
+ * `autoStopAfter` and it will manage the rest.
+ *
+ * Some details:
+ *
+ * - `autostopAfter` can be `0` if you want to stop sync as soon as it's done.
+ *
+ * - `autostopAfter` can be `Infinity` if you never want to auto-stop.
+ *
+ * - Changing `autostopAfter` does not restart the timer. For example, imagine
+ *   `autostopAfter` is `100` and sync completes at time `T`. If you change
+ *   `autostopAfter` to `200` at time `T < 100`, the timer will fire at time
+ *   `T + 200`.
+ */
+export default class Autostopper {
```

Comment:
I found the auto-stop logic fairly complex and decided to put it into its own (unit tested) class.